cmd/web-handlers.go

```diff
@@ -1332,44 +1331,20 @@ func (web *webAPIHandlers) DownloadZip(w http.ResponseWriter, r *http.Request) {
 				UncompressedSize64: uint64(length),
 				UncompressedSize:   uint32(length),
 			}
-			zipWriter, err := archive.CreateHeader(header)
+			if hasStringSuffixInSlice(gr.ObjInfo.Name, standardExcludeCompressExtensions) || hasPattern(standardExcludeCompressContentTypes, gr.ObjInfo.ContentType) {
+				// We strictly disable compression for standard extensions/content-types.
+				header.Method = zip.Store
+			}
+			writer, err := archive.CreateHeader(header)
```

Comment:
Shouldn't we decompress the compressed snappy stream and then copy it to this zip writer?

Reply:
This is the reason for the error above. We are directly passing the compressed content to the response writer. Thus producing a mismatch in size

Reply:
This is a bug fix for the existing code. Data is attempted to be decompressed twice, and this functionality does not work.

Call chain:
* [objectAPI.GetObjectNInfo](https://github.com/minio/minio/blob/f80508e1a3a567fd884d4f54ad92ad5188cdb4d3/cmd/web-handlers.go#L1313)
* xl: -> [NewGetObjectReader](https://github.com/minio/minio/blob/53e4887e02b94ed37dea8c5173330f17548fe122/cmd/xl-v1-object.go#L166)
* -> does decompression: https://github.com/minio/minio/blob/9eb8ef17e6a5a1ccbff53de9e274f0d64d812736/cmd/object-api-utils.go#L596

So for xl, we should receive a decompressed stream.

Same for FS, also using [NewGetObjectReader](https://github.com/minio/minio/blob/b52a3e523cc3c4debc0ea2f86386377df5355c81/cmd/fs-v1.go#L529).

I have confirmed that the current implementation is broken and ZIP download doesn't work on FS/XL. But there could be other backend implementations that rely on the broken functionality.

Reply:
yes @klauspost you're correct

Reply:
@Praveenrajmani But the error is concerning enough that I dont want to merge right now. Lets see if the error is consistent. 

It could be a single implementation that has specific problems or something else, so I will see if I can reproduce and find the cause.

Reply:
I think this would fix it. The `info.Size` was pointing the compressed size.

```
if info.IsCompressed() {
	// Read the decompressed size from the meta.json.
	actualSize = info.GetActualSize()
	// Set the info.Size to the actualSize.
	info.Size = actualSize
}
```

Reply:
Yeah, even though it only seems to be used for reporting, we probably should. Pushing a change shortly.